### PR TITLE
[FIX] mail: fix handling of lastVisibleMessage in channel seen condition

### PR DESCRIPTION
**Pull request by xdu-odoo**
also fix a problem with composer.hasFocus field not always correctly
updated (e.g. when switching from a channel to another in discuss)

task-2278915

### CODE REVIEW

**Comment by xdu-odoo** [comment]
see https://github.com/odoo/odoo/pull/59653 for origin story

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 1]
```diff
@@ -4206,6 +4206,125 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
     );
 });
 
+QUnit.test('mark channel as seen if last message visible when switching channels [REQUIRE FOCUS]', async function (assert) {
```
Comment (alexkuhn) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4209]: _"[...] if last message **is** visible [...]"_

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 2]
```diff
@@ -4206,6 +4206,125 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
...
+    assert.containsOnce(
+        document.body,
+        `.o_DiscussSidebar_item[data-thread-local-id="${
+            this.env.models['mail.thread'].findFromIdentifyingData({
+                id: 11,
+                model: 'mail.channel',
+            }).localId
+        }"]`,
+        "should have discuss sidebar item with the channel #11"
+    );
```
Comment (alexkuhn) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4247]: Assertions already covered by other(s) test(s).

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 3]
```diff
@@ -4206,6 +4206,125 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
...
+            .o_DiscussSidebar_item[data-thread-local-id="${
+                this.env.models['mail.thread'].findFromIdentifyingData({
+                    id: 11,
+                    model: 'mail.channel',
+                }).localId
+            }"]
+        `),
+        'o-unread',
+        "sidebar item of channel #11 should be unread"
+    );
```
Comment (alexkuhn) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4271]: Not sure whether it's covered by other(s) test(s), but should be dedicated test.

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 4]
```diff
@@ -4206,6 +4206,125 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
...
+            .o_DiscussSidebar_item[data-thread-local-id="${
+                this.env.models['mail.thread'].findFromIdentifyingData({
+                    id: 10,
+                    model: 'mail.channel',
+                }).localId
+            }"]
+        `),
+        'o-unread',
+        "sidebar item of channel ID 10 should not longer be unread"
+    );
```
Comment (alexkuhn) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4325]: Considering name of test, only the last assertion is relevant.

File: addons/mail/static/src/models/thread_view/thread_view.js [hunk 5]
```diff
@@ -113,6 +113,16 @@ function factory(dependencies) {
             }
         }
 
+        /**
+         * Last visible message is scoped to the thread cache, it's needed to be reseted when
```
Comment (alexkuhn) [inline addons/mail/static/src/models/thread_view/thread_view.js:117]: - Split "what" and "why" in 2 sentences. Basically replace comma by dot.
- _"reseted"_ => _"reset"_

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 3 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4271]: I don't agree, the intent of the test is to check that the switching make the channel becomes read so it needs to check it's unread before doing the action. Otherwise the test is not robust enough if something else before the action made it unread.

Comment (alexkuhn) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4271]: Problem with having assertions before is that test is less maintainable overall. It feels right but it isn't, that an illusion. You think that making test crash if premises aren't respected is good, but in the end it adds more noise in QUnit test crash and also makes test much longer than it should.

The purpose of **_this_** test is to check that is marks thread as read on switch when last message is visible. It's fine if it passes due to thread being already marked as read initially: another test should cover that exact scenario.

Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4271]: Still not totally convinced, but fine

Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4271]: Do we have this other test right now, and if so which one is it?

I feel like if we really fully want to go this way, we NEED to be able to express test dependencies, and if only manually with comments.

File: addons/mail/static/src/components/composer_text_input/composer_text_input.js [hunk 6]
```diff
@@ -153,6 +153,12 @@ class ComposerTextInput extends Component {
      * @private
      */
     _update() {
+        if (document.activeElement === this._textareaRef.el && !this.composer.hasFocus) {
+            // Needed to ensure the right composer is focused when the composer changes.
```
Comment (seb-odoo) [inline addons/mail/static/src/components/composer_text_input/composer_text_input.js:157]: This needs clarification. I think it can more simply be expressed by "Ensure focus state in composer record is kept in sync with actual focus in DOM. Indeed OWL might re-use the same component for displaying a different record."

But now that I write it, I think this is actually a mistake. I don't think the focus should be automatically kept like that on switching thread/composer. There is a "re-focus" because we do call "focus" when opening a new thread, but this shouldn't be done automatically at component level (what if we didn't want the focus in models??).

Can we find a way (t-key?) to instruct OWL to destroy it/remove its focus by default? Or do we have to live with this?

File: addons/mail/static/src/components/composer_text_input/composer_text_input.js [hunk 7]
```diff
@@ -179,14 +185,7 @@ class ComposerTextInput extends Component {
      * @private
      */
     _onFocusinTextarea() {
-        this.composer.update({ hasFocus: true });
-    }
-
-    /**
-     * @private
-     */
-    _onFocusoutTextarea() {
```
Comment (seb-odoo) [inline addons/mail/static/src/components/composer_text_input/composer_text_input.js:188]: Why remove focusout management? On clicking anywhere else on the interface than on the thread, we would like the composer to be marked as not focused no? Is that handled somewhere else I'm missing?

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 8]
```diff
@@ -4206,6 +4206,67 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
     );
 });
 
+QUnit.test('mark channel as seen if last message is visible when switching channels [REQUIRE FOCUS]', async function (assert) {
+    assert.expect(1);
+
+    // channel expected to be found in the sidebar, with the expected message_unread_counter
+    // and a random unique id that will be referenced in the test
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4213]: You can remove those comments, we knew it was a bad idea even when I added them, they are just too verbose and makes reading the test more complicated than just trying to figure out what the variables are.

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 9]
```diff
@@ -4206,6 +4206,67 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
...
+            .o_DiscussSidebar_item[data-thread-local-id="${
+                this.env.models['mail.thread'].findFromIdentifyingData({
+                    id: 10,
+                    model: 'mail.channel',
+                }).localId
+            }"]
+        `),
+        'o-unread',
+        "sidebar item of channel ID 10 should not longer be unread"
+    );
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4267]: Do we really need to do 10 -> 11 -> 10 (check 10) in this test to make sense?
Can't we just 10 -> 11 (check 11)?

Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4267]: It's not clear to me while we do all of those clicking around, and there's no comment or assertion to help.

File: addons/mail/static/src/components/message_list/message_list.js [hunk 10]
```diff
@@ -395,8 +395,8 @@ class MessageList extends Component {
             return;
         }
         if (
-            this.threadView.lastVisibleMessage &&
-            (message.id < this.threadView.lastVisibleMessage.id)
+            this.threadView.lastVisibleMessageOfThreadCache &&
+            (message.id < this.threadView.lastVisibleMessageOfThreadCache.id)
```
Comment (seb-odoo) [inline addons/mail/static/src/components/message_list/message_list.js:399]: This was a typo?

Comment (seb-odoo) [reply addons/mail/static/src/components/message_list/message_list.js:399]: Nvm js field rename in stable.
Could have kept the previous name and give a proper name in master.

But I guess since we are only 2 weeks in this is acceptable... Wouldn't do this in 3 months for sure though.

File: addons/mail/static/src/models/thread_view/thread_view.js [hunk 11]
```diff
@@ -217,7 +227,12 @@ function factory(dependencies) {
          * Most recent message in this ThreadView that has been shown to the
          * current partner.
          */
-        lastVisibleMessage: many2one('mail.message'),
+        lastVisibleMessageOfThreadCache: many2one('mail.message', {
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread_view/thread_view.js:230]: Actually I don't even like the new name.

This implies the "last visible message" is a thread cache notion, except it isn't the case, it's purely a view notion that is correctly computed here.

I think `lastVisibleMessage` was a better name. It's obvious in this context of thread view this is always referencing it's currently displayed thread cache.

Comment (seb-odoo) [reply addons/mail/static/src/models/thread_view/thread_view.js:230]: So for this reason + change in stable + minimal diff, I think name should be reverted.

File: addons/mail/static/src/components/composer_text_input/composer_text_input.js (hunk 6 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/composer_text_input/composer_text_input.js:157]: I don't know, I don't know OWL deep enough to answer your last question...
I'll just change the comment for now.

File: addons/mail/static/src/components/composer_text_input/composer_text_input.js (hunk 7 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/composer_text_input/composer_text_input.js:188]: I think it's a mistake from me, as in my uses cases the composer was focused out by the new focus method

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 9 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4267]: We do 11 -> 10, no channel is selected initially so we need to enable 11 and then 10

File: addons/mail/static/src/models/thread_view/thread_view.js (hunk 11 above)
Comment (xdu-odoo) [reply addons/mail/static/src/models/thread_view/thread_view.js:230]: @alexkuhn > @seb-odoo do not agree with your suggestion to change that field name, do I revert or can you convince him ?

Comment (alexkuhn) [reply addons/mail/static/src/models/thread_view/thread_view.js:230]: `lastVisibleMessageWithThreadCache`

Comment (alexkuhn) [reply addons/mail/static/src/models/thread_view/thread_view.js:230]: > I think `lastVisibleMessage` was a better name. It's obvious in this context of thread view this is always referencing it's currently displayed thread cache.

It's not obvious, notably it didn't make sense to reset this field on thread cache change. That's why `lastVisibleMessage` needs renaming so that it's more explicitly contextual to thread at hand.

Comment (xdu-odoo) [reply addons/mail/static/src/models/thread_view/thread_view.js:230]: @seb-odoo  > "La parole est à la défense"

Comment (seb-odoo) [reply addons/mail/static/src/models/thread_view/thread_view.js:230]: We shouldn't do rename of this field in stable if the rename is not necessary for the fix to work. For example on the transient PR it made sense to use another name because we changed the meaning of the field. In this case we don't want to change its meaning, the rename would just clarify its current meaning. The fix is to update the field correctly, not to change its meaning.

As for what to do in master, I would suggest the field be renamed with a `threadCache` prefix because that's the convention in that model for other such fields (and if we don't like that convention we can reconsider and rename all others). Note that whatever is done for this field should also be done for other thread cache fields in this thread view model, including `checkedMessages`, `isLoading`, `lastMessage` (should have "thread" not cache in name), `lastVisibleMessage`, `messages`, `uncheckedMessages`.

Comment (alexkuhn) [reply addons/mail/static/src/models/thread_view/thread_view.js:230]: > [...] on the transient PR it made sense to use another name because we changed the meaning of the field. In this case we don't want to change its meaning, the rename would just clarify its current meaning. The fix is to update the field correctly, not to change its meaning.

Motivation of field name change with transient PR is exactly the same as here: name is too loose with how we use the field, and the root cause is field name being not clear enough. This is the same here. Both cases are not drastic change of meaning: they are renamed to be more focused on what they currently do and what they mean, to avoid confusion.

The rename of fields is a fix, because without rename the field is badly named. This is especially true in the context of this PR. I wouldn't have mind not renaming the field without this PR, but with this PR this is self-evident.

> As for what to do in master, I would suggest the field be renamed with a `threadCache` prefix because that's the convention in that model for other such fields (and if we don't like that convention we can reconsider and rename all others). Note that whatever is done for this field should also be done for other thread cache fields in this thread view model, including `checkedMessages`, `isLoading`, `lastMessage` (should have "thread" not cache in name), `lastVisibleMessage`, `messages`, `uncheckedMessages`.

Agree that fields that are contextual to thread cache should have it explicit in name. Not super fond of suggested convention to prefix with `threadCache`, because will make code more cumbersome to read and understand what such fields do just by their name, but it's more acceptable having it than none like at present.

I haven't given too much thought, but maybe modeling could have some improvements to make it more obvious for these fields to have a lifetime dictate by thread cache at hand? Something like having a sub-model of `mail.thread_view` that is recreate on change of targeted thread cache? Basically would solve long naming of such fields and simplify the somewhat "short-lived" aspect of these fields (and also group all of them together).

Comment (seb-odoo) [reply addons/mail/static/src/models/thread_view/thread_view.js:230]: I'm not comfortable with renaming this field (or any other field) in stable. I've seen other fixes use it and it can cause issues if we are not very careful. If we least we knew what to name it, I would suggest to create the new field and keep this one as a deprecated related field to the other (for stable reasons) but we don't even agree on what would be its ideal new name. Just add comment to field to clarify it's scope, it should be more than enough to avoid issue for now. I'm still open to reconsider this name on master of course.

And actually about last paragraph this might be the solution, this is the same "problem" for scroll position btw of which we stored in the view with thread cache as key (and I already expressed that for me whenever a field is storing an array or an object, it should probably be a relation instead) so I guess making an intermediary model for those would indeed make sense.

Comment (alexkuhn) [reply addons/mail/static/src/models/thread_view/thread_view.js:230]: Ok for not changing name in 14.0, adding context of use as comment of field, and making change of field name in master. Let's aim for very minimal changes in stable versions ;)

File: addons/mail/static/src/components/composer_text_input/composer_text_input.xml [hunk 12]
```diff
@@ -10,7 +10,7 @@
                         isBelow="props.hasMentionSuggestionsBelowPosition"
                     />
                 </t>
-                <textarea class="o_ComposerTextInput_textarea" t-att-class="{ 'o-composer-is-compact': props.isCompact }" t-esc="composer.textInputContent" t-att-placeholder="textareaPlaceholder" t-on-focusin="_onFocusinTextarea" t-on-focusout="_onFocusoutTextarea" t-on-keydown="_onKeydownTextarea" t-on-keyup="_onKeyupTextarea" t-on-input="_onInputTextarea" t-ref="textarea"/>
+                <textarea class="o_ComposerTextInput_textarea" t-att-class="{ 'o-composer-is-compact': props.isCompact }" t-esc="composer.textInputContent" t-att-placeholder="textareaPlaceholder" t-on-focusin="_onFocusinTextarea" t-on-keydown="_onKeydownTextarea" t-on-keyup="_onKeyupTextarea" t-on-input="_onInputTextarea" t-ref="textarea"/>
```
Comment (seb-odoo) [inline addons/mail/static/src/components/composer_text_input/composer_text_input.xml:13]: focusout is still removed here, shouldn't we revert this too?

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 13]
```diff
@@ -4206,6 +4206,65 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
...
+        document.querySelector(`
+            .o_DiscussSidebar_item[data-thread-local-id="${
+                this.env.models['mail.thread'].findFromIdentifyingData({
+                    id: 10,
+                    model: 'mail.channel',
+                }).localId
+            }"]
+        `),
+        'o-unread',
+        "sidebar item of channel ID 10 should not longer be unread"
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4264]: "no longer"

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 14]
```diff
@@ -4206,6 +4206,65 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
     );
 });
 
+QUnit.test('mark channel as seen if last message is visible when switching channels [REQUIRE FOCUS]', async function (assert) {
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4209]: I think we also need to add "... even when the previous channel had a more recent last message than the current channel"
because this was actually the bug. In the normal case as described by current title it already worked before.

File: addons/mail/static/src/models/thread_view/thread_view.js [hunk 15]
```diff
@@ -217,7 +227,12 @@ function factory(dependencies) {
          * Most recent message in this ThreadView that has been shown to the
          * current partner.
          */
-        lastVisibleMessage: many2one('mail.message'),
+        lastVisibleMessage: many2one('mail.message', {
+            compute: '_computeLastVisibleMessage',
+            dependencies: [
+                'threadCache',
+            ]
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread_view/thread_view.js:234]: `,`

File: addons/mail/static/src/models/thread_view/thread_view.js [hunk 16]
```diff
@@ -217,7 +227,12 @@ function factory(dependencies) {
          * Most recent message in this ThreadView that has been shown to the
          * current partner.
          */
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread_view/thread_view.js:229]: PR looks much nicer without the rename, but as discussed, add comment about scope of field into the field docstring too.

File: addons/mail/static/src/models/composer/composer.js [hunk 17]
```diff
@@ -102,6 +102,19 @@ function factory(dependencies) {
             }
         }
 
+        /**
+         * Focus this composer and remove focus from all others.
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer/composer.js:106]: please explain why (focus is a global concern and it makes no sense to have multiple composers focused at the same time)

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 14 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4209]: Shouldn't we actually have 2 tests here? 1 for the focus issue, and 1 for the last message issue?

**Review by seb-odoo — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 14 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4209]: I don't know how you could test the focus issue ? By the way, the .focus() is no more needed in this test, we could remove it

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 3 above)
Comment (alexkuhn) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: > I feel like if we really fully want to go this way, we NEED to be able to express test dependencies, even if only manually with comments.

Really annoying QUnit lacking way to express dependencies among tests... To the point I would have liked not using QUnit at all and just have our own tools for unit testing...

Expressing dependencies in comment, considering how weak it is (not visible in QUnit UI, debug requires reading actual comment, likely comment becomes outdated, etc.), I prefer providing expected working behaviour in plain English rather than referencing other test. I know it sucks, but that's the best I can find without actual support of test dependencies...

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 14 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4209]: Definitely remove the focus from this test if it's no longer needed, and update title as suggested above.

As for how to test the focus issue... Well the same way you detected it in the first place, just gotta reproduce that flux in a test.

File: addons/mail/static/src/models/thread_view/thread_view.js [hunk 18]
```diff
@@ -113,6 +113,16 @@ function factory(dependencies) {
...
 
+        /**
+         * Last visible message is scoped to the thread cache.
+         * Needed to be reset when threadCache changes.
+         *
+         * @private
+         */
+        _computeLastVisibleMessage() {
+            return [['unlink']];
+        }
```
Comment (alexkuhn) [inline addons/mail/static/src/models/thread_view/thread_view.js:124]: I really don't like this... This is just a hidden `updateAfter()`. Also problematic when removing dependencies in definition of computed fields.

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 3 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: Well using phrases to describe initial state is more complicated and less robust than just keeping initial assertions I think...

Only dependencies avoids duplicating assertion/explanation.

File: addons/mail/static/src/models/thread_view/thread_view.js (hunk 18 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/thread_view/thread_view.js:124]: I wouldn't worry about it in stable.
We can/will be able to fix this only if/when we make thread_view_cache model as you suggested in another message.
This is very similar to when we had activities in chatter, the only way to get rid of this crap was to move them to another proper model.

Comment (seb-odoo) [reply addons/mail/static/src/models/thread_view/thread_view.js:124]: Note that removing dependencies is a general problem you will have with all "on change" computes, not just this one.

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 3 above)
Comment (alexkuhn) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: Honestly can't find decent replacement for lack of dependencies... Putting assertions in premise was our choice since not long ago, but these unit tests become hard to maintain, even when they are nicely written.

Of course they are still much better than JS tests outside of discuss, it's night and day... but the increase of maintenance cost of this way of writing test is too significant for my taste. This problem is crescendo, so better improve now than later.

File: addons/mail/static/src/models/thread_view/thread_view.js (hunk 18 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/thread_view/thread_view.js:124]: > Note that removing dependencies is a general problem you will have with all "on change" computes, not just this one.

Good point. Reason to make dummy field `onX` to explicitly reset `lastVisibleMessage`.
"on change" computes are in a category of their own, I prefer dedicate struggling with them without other conventional computes.

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 3 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: I agree for new solution, and I think it's acceptable and better than alternatives too (I was never in favor of copy/pasting all those assertions in the first place anyway), just have to keep in mind we want dependencies or at least ability to execute tests in certain order.

Considering everything we are writing ourselves at Odoo, rewriting a test framework for that purpose doesn't even look like a big task to be honest.

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 14 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js:4209]: Yeah but the problem was in model layer, so difficult to check anything in the UI ..

File: addons/mail/static/src/models/thread_view/thread_view.js (hunk 18 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/thread_view/thread_view.js:124]: So is that a request for @xdu-odoo to move this compute to a separate `on` field?

Comment (alexkuhn) [reply addons/mail/static/src/models/thread_view/thread_view.js:124]: Yes, sorry if it wasn't clear enough.

@xdu-odoo Don't make `lastVisibleMessage` field computed: instead, make a dummy computed field `onChangeThreadCache` and explicitly unlink `lastVisibleMessage` in compute method.

File: addons/mail/static/src/components/composer_text_input/composer_text_input.js (hunk 6 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/composer_text_input/composer_text_input.js]: Just try with a `t-key` based on `composer.localId` I guess and if that doesn't work then it's probably not possible?

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 9 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: Can't we use activeId to select 11 by default to avoid this step and then directly switch to 10?

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 14 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: I mean the fix of focus that we do in `_update` in this PR, this one is in a component at least so there must be a way?

Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: in the UI nothing changes really.. in fact the both fix are needed to fix the found problem : we need the focus to be "redispatched" when switching channels AND we need the lastMessage to be bound the thread shown as otherwise the markAsSeen is not called as the shown message is older than the last notified as seen or the composer is already focused

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 9 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: I don't know, do you have an example where it's used ?

Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: In business code it's just setting `initActiveId` but it appears we never test it. I guess this would require changing `start` helper to allow passing this value at create of discuss, so that, when `openInitThread` is called at init, it would read this value instead of "no thread".

Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: I found it actually:

```    await this.start({
        discuss: {
            params: {
                default_active_id: 'mail.box_moderation',
            },
        },
    });

```

Comment (alexkuhn) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: By the way, don't use `default_active_id`. Should only be used as data of the action.
Prefer `discuss.context.active_id` instead.

_(yes, we made this small mistake in most of our tests)_

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 14 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: That means if we wanted to test the focus case we would have to write the exact same test except the thread to which we switch into has a more recent message instead of least recent? But this is the standard case, we never noticed that issue in practice I think? Was the focus issue only manifesting in the current test, or also when testing manually?

Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: Only in test, because when you click in real situation, the focus is also changed so the problem was not showing up, but in test you had to focus and then click ... By the way the t-key seems to work too, so maybe we don't need to test this anymore as it's strongly related to test conditions ?

File: addons/mail/static/src/components/composer_text_input/composer_text_input.js (hunk 6 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/composer_text_input/composer_text_input.js]: Seems to work

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 3 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: anything to change here ?

File: addons/mail/static/src/components/composer/composer.xml [hunk 19]
```diff
@@ -83,6 +83,7 @@
                         t-on-o-composer-suggestion-clicked="_onComposerSuggestionClicked"
                         t-on-o-composer-text-input-send-shortcut="_onComposerTextInputSendShortcut"
                         t-on-paste="_onPasteTextInput"
+                        t-key="composer.localId"
```
Comment (seb-odoo) [inline addons/mail/static/src/components/composer/composer.xml:86]: Maybe this should be on `<Composer/>` itself it that also works? Actually I'm really wondering if we shouldn't do this fix for all of our components that are related to models.

File: addons/mail/static/src/models/composer/composer.js [hunk 20]
```diff
@@ -102,6 +102,21 @@ function factory(dependencies) {
...
+         * Focus is a global concern, it makes no sens to have multiple composers focused at the
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer/composer.js:107]: typo sens

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 3 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: let's consider it fine for now, just to keep in mind

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js (hunk 14 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/tests/discuss_tests.js]: Ok to just keep this test.

File: addons/mail/static/src/components/composer/composer.xml (hunk 19 above)
Comment (xdu-odoo) [reply addons/mail/static/src/components/composer/composer.xml:86]: on the main div of composer.xml on or each <Composer> node ?

Comment (seb-odoo) [reply addons/mail/static/src/components/composer/composer.xml:86]: That's a good question actually. I was thinking on each node, to ensure we actually get a new component when we change record. But maybe just on the main div would make less noise everywhere. I think the end result will be the same for focus. But it might be different for other cases (e.g. variables we store on the component itself, and especially when we do processing in `_update` with check depending on previous/new record being displayed, those could be removed if we made that change I think).

Comment (seb-odoo) [reply addons/mail/static/src/components/composer/composer.xml:86]: Considering this is a fix in stable, let's just keep it here on textinput where we have an issue. We can try more drastic changes in master. (I think it would be beneficial to investigate this asap)

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 21]
```diff
@@ -4206,6 +4206,53 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
...
+        body: 'newest message',
+        channel_ids: [11],
+        id: 11,
+    });
+    await this.start({
+        discuss: {
+            context: {
+                active_id:  'mail.channel_11',
+            }
+        }
```
Comment (seb-odoo) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4230]: `,` `,`

File: addons/mail/static/src/models/composer/composer.js (hunk 20 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/composer/composer.js:107]: Don't forget to push the change

**Review by seb-odoo — APPROVED** [review]
Last detail but this should be good now, @alexkuhn final check?

File: addons/mail/static/src/models/thread_view/thread_view.js [hunk 22]
```diff
@@ -210,6 +211,9 @@ function factory(dependencies) {
          * hint `message-received`.
          */
         hasAutoScrollOnMessageReceived: attr(),
+        /**
+         * Last message in the context of the currently displayed thread cache.
+         */
```
Comment (alexkuhn) [inline addons/mail/static/src/models/thread_view/thread_view.js:216]: Similar aspect with `lastVisibleMessage` (dependent on current thread cache).
Would update comment of `lastVisibleMessage` too, because in context of PR was wondering why `unlink` on thread cache changed, which introduced the debate of rename, hence need clarification.

File: addons/mail/static/src/components/composer/composer.xml (hunk 19 above)
Comment (alexkuhn) [reply addons/mail/static/src/components/composer/composer.xml:86]: I haven't caught the whole discussion, but I assume something was off with reusing `Composer` component on change of thread cache of thread view, so this enforces change of component.

Ok for this hack. In theory this issue should not exist (stateless components), but whatever.

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 23]
```diff
@@ -4206,6 +4206,53 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
     );
 });
 
+QUnit.test('mark channel as seen if last message is visible when switching channels even when the previous channel had a more recent last message than the current channel', async function (assert) {
```
Comment (alexkuhn) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:4209]: Would remove "even" in the test name. It's confusing when the test actual test specifically this scenario, nothing else.

**Review by alexkuhn — CHANGES_REQUESTED** [review]
Minus small comments, LGTM.

**Review by alexkuhn — APPROVED** [review]
@robodoo r+
